File: android/app/src/main/java/com/vietbridge/ai/ui/auth/AuthViewModel.kt
package com.vietbridge.ai.ui.auth

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.vietbridge.ai.data.local.TokenManager
import com.vietbridge.ai.data.model.LoginRequest
import com.vietbridge.ai.data.model.RegisterRequest
import com.vietbridge.ai.data.model.User
import com.vietbridge.ai.data.remote.ApiClient
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class AuthState(
    val isLoading: Boolean = true,
    val isAuthenticated: Boolean = false,
    val isGuest: Boolean = false,
    val user: User? = null,
    val error: String? = null,
    val success: String? = null,
)

class AuthViewModel(private val tokenManager: TokenManager) : ViewModel() {

    private val _state = MutableStateFlow(AuthState())
    val state: StateFlow<AuthState> = _state.asStateFlow()

    init {
        ApiClient.init(tokenManager)
        checkAuth()
    }

    private fun checkAuth() {
        if (tokenManager.token == null) {
            // No token → enter as guest (skip login screen)
            _state.update { it.copy(isLoading = false, isAuthenticated = true, isGuest = true) }
            return
        }
        viewModelScope.launch {
            try {
                val user = ApiClient.api.me()
                _state.update { it.copy(isLoading = false, isAuthenticated = true, isGuest = false, user = user) }
            } catch (_: Exception) {
                tokenManager.clear()
                // Auth failed → still enter as guest
                _state.update { it.copy(isLoading = false, isAuthenticated = true, isGuest = true) }
            }
        }
    }

    fun login(email: String, password: String) {
        _state.update { it.copy(isLoading = true, error = null) }
        viewModelScope.launch {
            try {
                val response = ApiClient.api.login(LoginRequest(email, password))
                tokenManager.token = response.token
                _state.update { it.copy(isLoading = false, isAuthenticated = true, isGuest = false, user = response.user) }
            } catch (e: Exception) {
                _state.update { it.copy(isLoading = false, error = e.message ?: "登录失败") }
            }
        }
    }

    fun register(name: String, email: String, password: String) {
        _state.update { it.copy(isLoading = true, error = null) }
        viewModelScope.launch {
            try {
                val response = ApiClient.api.register(RegisterRequest(name, email, password))
                tokenManager.token = response.token
                _state.update { it.copy(isLoading = false, isAuthenticated = true, isGuest = false, user = response.user) }
            } catch (e: Exception) {
                _state.update { it.copy(isLoading = false, error = e.message ?: "注册失败") }
            }
        }
    }

    fun forgotPassword(email: String) {
        _state.update { it.copy(isLoading = true, error = null, success = null) }
        viewModelScope.launch {
            try {
                ApiClient.api.forgotPassword(com.vietbridge.ai.data.model.ForgotPasswordRequest(email))
                _state.update { it.copy(isLoading = false, success = "重置链接已发送到您的邮箱") }
            } catch (e: Exception) {
                _state.update { it.copy(isLoading = false, error = e.message ?: "发送失败") }
            }
        }
    }

    fun signOut() {
        tokenManager.clear()
        _state.update { AuthState(isLoading = false, isAuthenticated = true, isGuest = true) }
    }

    fun clearMessages() {
        _state.update { it.copy(error = null, success = null) }
    }

    class Factory(private val tokenManager: TokenManager) : ViewModelProvider.Factory {
        @Suppress("UNCHECKED_CAST")
        override fun <T : ViewModel> create(modelClass: Class<T>): T = AuthViewModel(tokenManager) as T
    }
}

File: ios/VietBridge/ViewModels/AuthViewModel.swift
// ============================================================================
// VietBridge AI — Auth View Model
// Manages login/register/forgot password state
// ============================================================================

import SwiftUI

@Observable
@MainActor
final class AuthViewModel {
    var email = ""
    var password = ""
    var name = ""
    var confirmPassword = ""
    var isLoading = false
    var errorMessage: String?
    var successMessage: String?
    var mode: AuthMode = .login

    enum AuthMode {
        case login
        case register
        case forgotPassword
    }

    private let authService = AuthService()

    func login(appState: AppState) async {
        guard validate() else { return }
        isLoading = true
        errorMessage = nil

        do {
            try await appState.login(email: email, password: password)
        } catch {
            errorMessage = error.localizedDescription
        }

        isLoading = false
    }

    func register(appState: AppState) async {
        guard validateRegister() else { return }
        isLoading = true
        errorMessage = nil

        do {
            try await appState.register(name: name, email: email, password: password)
        } catch {
            errorMessage = error.localizedDescription
        }

        isLoading = false
    }

    func forgotPassword() async {
        guard !email.isEmpty else {
            errorMessage = "请输入邮箱"
            return
        }
        isLoading = true
        errorMessage = nil

        do {
            try await authService.forgotPassword(email: email)
            successMessage = "重置链接已发送到您的邮箱"
        } catch {
            errorMessage = error.localizedDescription
        }

        isLoading = false
    }

    func clearMessages() {
        errorMessage = nil
        successMessage = nil
    }

    private func validate() -> Bool {
        if email.isEmpty || password.isEmpty {
            errorMessage = "请填写邮箱和密码"
            return false
        }
        return true
    }

    private func validateRegister() -> Bool {
        if name.isEmpty {
            errorMessage = "请填写姓名"
            return false
        }
        if email.isEmpty || password.isEmpty {
            errorMessage = "请填写邮箱和密码"
            return false
        }
        if password.count < 6 {
            errorMessage = "密码至少6个字符"
            return false
        }
        if password != confirmPassword {
            errorMessage = "两次密码输入不一致"
            return false
        }
        return true
    }
}
